### PR TITLE
[Merged by Bors] - feat: biSup, biInf, biUnion and biInter in NoMaxOrder

### PR DESCRIPTION
**Pull request by scholzhannah**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by eric-wieser — COMMENTED** [review]
Could you add the `NoMinOrder` versions too, and perhaps also the `le` versions? (so 4x what you have right now!)

File: Mathlib/Data/Set/Lattice.lean [hunk 1]
```diff
@@ -1173,6 +1173,30 @@ theorem union_distrib_iInter_right (s : ι → Set α) (t : Set α) : (⋂ i, s
 theorem union_distrib_iInter₂_right (s : ∀ i, κ i → Set α) (t : Set α) :
     (⋂ (i) (j), s i j) ∪ t = ⋂ (i) (j), s i j ∪ t := by simp_rw [union_distrib_iInter_right]
 
+lemma biUnion_lt_eq_iUnion [LT α] [NoMaxOrder α] {s : α → Set β} :
+    ⋃ (n) (m < n), s m  = ⋃ n, s n := biSup_lt_eq_iSup
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Set/Lattice.lean:1177]:
```suggestion
    ⋃ (n) (m < n), s m = ⋃ n, s n := biSup_lt_eq_iSup
```
and in all the other lemmas in this file

File: Mathlib/Order/CompleteLattice.lean [hunk 2]
```diff
@@ -1072,6 +1072,62 @@ theorem inf_biInf {p : ι → Prop} {f : ∀ i, p i → α} {a : α} (h : ∃ i,
...
+  · exact iSup_le fun _ ↦ iSup_le fun _ ↦ iSup_le fun _ ↦ le_iSup _ _
+  · exact iSup_le fun j ↦ le_iSup_of_le j (le_iSup_of_le j (le_iSup_of_le (le_refl _) (le_refl _)))
+
+lemma biInf_lt_eq_iInf {ι : Type*} [LT ι] [NoMaxOrder ι] {f : ι → α} :
+    ⨅ (i) (j < i), f j = ⨅ i, f i := by
+  apply le_antisymm
+  · apply le_iInf fun j ↦ ?_
+    obtain ⟨i, jlt⟩ := exists_gt j
+    apply iInf_le_of_le i (iInf_le_of_le j (iInf_le_of_le jlt (le_refl _)))
+  · exact le_iInf fun _ ↦ le_iInf fun _ ↦ le_iInf fun _ ↦ iInf_le _ _
```
Comment (fpvandoorn) [inline Mathlib/Order/CompleteLattice.lean:1095]:
```suggestion
lemma biInf_lt_eq_iInf {ι : Type*} [LT ι] [NoMaxOrder ι] {f : ι → α} :
    ⨅ (i) (j < i), f j = ⨅ i, f i := 
  biSup_lt_eq_iSup (α := αᵒᵈ)
```
You don't have to repeat the proof, but can say "this is the same statement, but with the order on `α` flipped" (`αᵒᵈ` is the same type as `α`, but with the dual order).

File: Mathlib/Order/CompleteLattice.lean [hunk 3]
```diff
@@ -1072,6 +1072,62 @@ theorem inf_biInf {p : ι → Prop} {f : ∀ i, p i → α} {a : α} (h : ∃ i,
...
+  · exact le_iInf fun j ↦ iInf_le_of_le j (iInf_le_of_le j (iInf_le_of_le (le_refl _) (le_refl _)))
+  · exact le_iInf fun _ ↦ le_iInf fun _ ↦ le_iInf fun _ ↦ iInf_le _ _
+
+lemma biSup_gt_eq_iSup {ι : Type*} [LT ι] [NoMinOrder ι] {f : ι → α} :
+    ⨆ (i) (j > i), f j = ⨆ i, f i := by
+  apply le_antisymm
+  · exact iSup_le fun _ ↦ iSup_le fun _ ↦ iSup_le fun _ ↦ le_iSup _ _
+  · apply iSup_le fun j ↦ ?_
+    obtain ⟨i, ltj⟩ := exists_lt j
+    exact le_iSup_of_le i (le_iSup_of_le j (le_iSup_of_le ltj (le_refl _)))
```
Comment (fpvandoorn) [inline Mathlib/Order/CompleteLattice.lean:1109]:
```suggestion
lemma biSup_gt_eq_iSup {ι : Type*} [LT ι] [NoMinOrder ι] {f : ι → α} :
    ⨆ (i) (j > i), f j = ⨆ i, f i := 
  biSup_lt_eq_iSup (ι := ιᵒᵈ)
```

**Review by fpvandoorn — COMMENTED** [review]
Please also do this for the remaining lemmas (where possible)

**Review by eric-wieser — APPROVED** [review]
bors merge

Thanks!

**Pull request closed without merging**